### PR TITLE
Apply Optimization in Index.GetOffset suggested by Levi

### PR DESCRIPTION
**Pull request by tarekgh**

### CODE REVIEW

**Comment by tarekgh** [comment]
@GrabYourPitchforks thanks for the suggestion.

**Comment by stephentoub** [comment]
Perf numbers?

**Comment by tarekgh** [comment]
CC @AndyAyersMS for the jit side.

I have run some perf measurements on this change and I am seeing some slight enhancement around 3%

**Before** 

```
        Method |     Mean |    Error |   StdDev |
-------------- |---------:|---------:|---------:|
 GetOffsetTest | 615.6 us | 12.57 us | 35.46 us |
```

**After**

```
        Method |     Mean |    Error |   StdDev |
-------------- |---------:|---------:|---------:|
 GetOffsetTest | 596.8 us | 11.90 us | 22.35 us |
```

But interesting enough, looks the JIT was applying some optimization before and this optimization is not there anymore after this change but we still seeing some slight perf enhancement. I had the test code like:

```C#
            Index index = new Index(1, true); // index from the end
        
            for (int j = 0; j < 2000000; j++)
            {
              int offset = index.GetOffset(100);
            }
```

before the change, I am seeing jit was producing the following code for GetOffsetTest

```
 System.Index.GetOffset(Int32)	
not     eax
sub     edx,eax
jmp     System_Private_CoreLib+0x3d2193
mov     edx,eax
mov     eax,edx
```

Somehow the check inside was eliminated. And after the change I am seeing the jitted code is 

```
 System.Index.GetOffset(Int32)	
mov     eax,dword ptr [rcx]
mov     ecx,eax
test    eax,eax
jge     System_Private_CoreLib!System.Index.GetOffset(Int32)$##6001330+0x11
lea     ecx,[rcx+rdx+1]
mov     eax,ecx
```

which is including the check but interesting we still seeing some slight perf improvement.

**Comment by AndyAyersMS** [comment]
Be careful how you benchmark. When I try I get identical code before & after. If I literally use your test code I get an empty loop:
```asm
G_M5602_IG03:
       FFC0                 inc      eax
       3D80841E00           cmp      eax, 0x1E8480
       7CF7                 jl       SHORT G_M5602_IG03
```
and if I consume the result of `GetOffset` (say accumulating it) I get something like this for both versions:
```asm
G_M5601_IG03:
       B963000000           mov      ecx, 99
       03C1                 add      eax, ecx
       FFC2                 inc      edx
       81FA80841E00         cmp      edx, 0x1E8480
       7CEF                 jl       SHORT G_M5601_IG03
```
Ideally you should benchmark this in context of how it is likely to be used. Indexes constructed locally with constant parameters will likely be completely optimized away.

**Comment by tarekgh** [comment]
@AndyAyersMS 

would the following test code be reasonable for the benchmark? 

```C#
        [Benchmark] public char GetOffsetTest()
        {
            int lastOffset = 0;
            char c = ' ';

            ReadOnlySpan<char> span = new string('a', 400).AsSpan();

            for (int j = 0; j < 2000000; j++)
            {
                Index index = new Index(j % span.Length, true);
                int offset = index.GetOffset(j * 2);
                if (offset < span.Length)
                {
                    c = span[offset];
                }
            }

            return c;
        }
```

I tried this code and I am still seeing some similar results 

**Before**

```

        Method |     Mean |     Error |    StdDev |
-------------- |---------:|----------:|----------:|
 GetOffsetTest | 7.900 ms | 0.1572 ms | 0.3973 ms |
 
 
 System.Index.GetOffset(Int32)	
not     eax
sub     edx,eax
jmp     System_Private_CoreLib!System.Index.GetOffset(Int32)$##60012CE+0x13
mov     edx,eax
mov     eax,edx


```


**After**

```
        Method |     Mean |     Error |    StdDev |
-------------- |---------:|----------:|----------:|
 GetOffsetTest | 6.534 ms | 0.1265 ms | 0.2723 ms |
 
 System.Index.GetOffset(Int32)	
mov     eax,dword ptr [rcx]
mov     ecx,eax
test    eax,eax
jge     System_Private_CoreLib!System.Index.GetOffset(Int32)$##6001330+0x11
lea     ecx,[rcx+rdx+1]
mov     eax,ecx

```

The perf enhancement with the fix is bit better though.

**Comment by AndyAyersMS** [comment]
It's somewhat better. I would use the index in the span index-based indexer (which calls `GetOffset`), since that's what most users will do. The `%` might be kind of expensive, so perhaps use something cheaper like `&`:

Something like:
```C#
    public static int Bench()
    {
        ReadOnlySpan<char> span = new string('a', 400).AsSpan();

        char result = ' ';
        
        for (int j = 0; j < 2000000; j++)
        {
            index = new Index(j & span.Length, true);
            result ^= span[index];
        }

        return result;
    }
```
I would also report a second pair of results for `new Index(..., false)`.

The change itself is a partial workaround for some of the issues #22130. You might want to explain in a comment why we need the `+1`.

**Comment by tarekgh** [comment]
Here is the result according to the suggested test by @AndyAyersMS 

**Before**
```
                      Method |      Mean |     Error |    StdDev |    Median |
---------------------------- |----------:|----------:|----------:|----------:|
   GetOffsetTestWithEndIndex | 12.707 ms | 0.2530 ms | 0.6393 ms | 12.507 ms |
 GetOffsetTestWithStartIndex |  9.978 ms | 0.1982 ms | 0.4709 ms |  9.926 ms |
 
```

**After**

```

                       Method |      Mean |     Error |    StdDev |
---------------------------- |----------:|----------:|----------:|
   GetOffsetTestWithEndIndex | 10.279 ms | 0.2796 ms | 0.7886 ms |
 GetOffsetTestWithStartIndex |  9.746 ms | 0.1912 ms | 0.2862 ms |

```

Which is almost 20% improvement in the Index from the end case and 2% in the Index from the start case.

I'll add the comment according as suggested too.

File: src/System.Private.CoreLib/shared/System/Index.cs [hunk 1]
```diff
@@ -103,13 +103,14 @@ public int Value
...
-            int offset;
-
+            int offset = _value;
             if (IsFromEnd)
-                offset = length - (~_value);
-            else
-                offset = _value;
-
+            {
+                // offset = length - value
```
Comment (GrabYourPitchforks) [inline src/System.Private.CoreLib/shared/System/Index.cs]:
```txt
offset = length - (~value)
offset = length + (~(~value) + 1)
offset = length + value + 1
```

**Pull request merged**